pkg/integrations/v2/app_o11y_receiver/handler/handler.go

```diff
@@ -0,0 +1,124 @@
+	exporterErrorsCollector *prometheus.CounterVec
+}
+
+// NewAppO11yHandler creates a new AppReceiver instance based on the given configuration
+func NewAppO11yHandler(conf config.AppO11yReceiverConfig, exporters []exporters.AppO11yReceiverExporter, reg *prometheus.Registry) AppO11yHandler {
+	var rateLimiter *ratelimiting.RateLimiter
+	if conf.RateLimiting.Enabled {
+		var rps float64
+		if conf.RateLimiting.RPS > 0 {
+			rps = conf.RateLimiting.RPS
+		}
+
+		var b int
+		if conf.RateLimiting.Burstiness > 0 {
+			b = conf.RateLimiting.Burstiness
+		}
+		rateLimiter = ratelimiting.NewRateLimiter(rps, b)
+	}
+
+	exporterErrorsCollector := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: utils.MetricsNamespace,
+		Subsystem: "exporter",
+		Name:      "errors",
+		Help:      "Total number of errors produced by a receiver exporter",
+	}, []string{"exporter"})
+
+	if reg != nil {
+		reg.MustRegister(exporterErrorsCollector)
+	}
+
+	return AppO11yHandler{
+		exporters:               exporters,
+		config:                  conf,
+		rateLimiter:             rateLimiter,
+		exporterErrorsCollector: exporterErrorsCollector,
+	}
+}
+
+// HTTPHandler is the http.Handler for the receiver. It will do the following
+// 0. Enable CORS for the configured hosts
+// 1. Check if the request should be rate limited
+// 2. Verify that the payload size is within limits
+// 3. Start two go routines for exporters processing and exporting data respectively
+// 4. Respond with 202 once all the work is done
+func (ar *AppO11yHandler) HTTPHandler(logger log.Logger) http.Handler {
+	var handler http.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Check rate limiting state
+		if ar.config.RateLimiting.Enabled && ar.rateLimiter.IsRateLimited() {
+			http.Error(w, http.StatusText(http.StatusTooManyRequests), http.StatusTooManyRequests)
+			return
+		}
+
+		// check API key if one is provided
+		if len(ar.config.APIKey) > 0 && subtle.ConstantTimeCompare([]byte(r.Header.Get("x-api-key")), []byte(ar.config.APIKey)) == 0 {
+			http.Error(w, "api key not provided or incorrect", http.StatusUnauthorized)
+			return
+		}
+
+		// Verify content length. We trust net/http to give us the correct number
+		if ar.config.MaxAllowedPayloadSize > 0 && r.ContentLength > ar.config.MaxAllowedPayloadSize {
+			http.Error(w, http.StatusText(http.StatusRequestEntityTooLarge), http.StatusRequestEntityTooLarge)
+			return
+		}
+
+		var p models.Payload
+		err := json.NewDecoder(r.Body).Decode(&p)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		var wg sync.WaitGroup
+
+		for _, exporter := range ar.exporters {
+			wg.Add(1)
+			go func(exp exporters.AppO11yReceiverExporter) {
+				defer wg.Done()
+				if err := exp.Export(r.Context(), p); err != nil {
+					level.Error(logger).Log("msg", "exporter error", "exporter", exp.Name(), "error", err)
+					ar.exporterErrorsCollector.WithLabelValues(exp.Name()).Inc()
```

Comment:
Should this cause a non-200 to be returned?

Reply:
Given that a) Partial success is possible, eg ingesting logs succeeds but traces fail b) Client shouldn't do anything if this fails, we decided to just return `202 Accepted `always. The thinking is that if at some point we need client to know the final outcome, we'd implement some sort of multiple result response